### PR TITLE
Avoid defensive copy in String.Concat(string[])

### PR DESCRIPTION
**Pull request by stephentoub**
Today the String.Concat(params string[]) implementation makes a defensive copy of the input string[] in order to avoid issues where another thread mutates the array concurrently with the concatenation.  Such a situation is possible but exceedingly rare, so we can redo the implementation to make it cheaper when there isn't concurrent mutation and more expensive when there is, rather than always having it be more expensive.

This commit changes string.Concat to optimistically assume there won't be such concurrent mutation.  It avoids allocating the defensive string[] array copy, and instead just proceeds to allocate a string of the right length and copy the input strings into it.  If along the way it discovers that something has changed such that the string lengths no longer add up to exactly what was precomputed, then it falls back to the original implementation.

Example microbenchmark:
```C#
using System;
using System.Diagnostics;

public class Program
{
    public static void Main()
    {
        string result;
        string[] inputs = new[] { "abcd", "efgh", "ijkl", "mnop", "qrst", "uvwx", "yz" };
        var sw = new Stopwatch();
        while (true)
        {
            int gen0 = GC.CollectionCount(0);
            sw.Restart();
            for (int i = 0; i < 20000000; i++) result = string.Concat(inputs);
            sw.Stop();
            Console.WriteLine($"{GC.CollectionCount(0) - gen0}: {sw.Elapsed.TotalSeconds}");
        }
    }
}
```
Before the change:
```
> corerun test.exe
1525: 2.0733829
1526: 2.0599474
1526: 2.0717786
1526: 2.0318797
^C
```
After the change:
```
> corerun test.exe
763: 1.4700695
762: 1.446919
763: 1.4581139
763: 1.4568889
^C
```

cc: @jkotas, @ellismg

### CODE REVIEW

File: src/mscorlib/src/System/String.cs [hunk 1]
```diff
@@ -3339,13 +3339,73 @@ public static String Concat(Object arg0, Object arg1, Object arg2, Object arg3,
...
+                }
+            }
+
+            // If we copied exactly the right amount, return the new string.  Otherwise,
+            // something changed concurrently to mutate the input array: fall back to
+            // doing the concatenation again, but this time with a defensive copy. This
+            // fall back should be extremely rare.
+            return copiedLength == totalLength ? 
+                result : 
+                ConcatWithDefensiveCopy(values);
```
Comment (jkotas) [inline src/mscorlib/src/System/String.cs]: We should not need to duplicate the implementation for defensive copy - it should be fine sufficient to just call the method recursively as `Contact(values.Clone())`.

Comment (sonergonul) [inline src/mscorlib/src/System/String.cs]: I think you mean `Concat `:)

Comment (jkotas) [inline src/mscorlib/src/System/String.cs]: Yes...

**Comment by jkotas** [comment]
Do we have tests in corefx to exercise the rare path?

File: src/mscorlib/src/System/String.cs (hunk 1 above)
Comment (stephentoub) [inline src/mscorlib/src/System/String.cs]: Good point, will change.

**Comment by stephentoub** [comment]
> Do we have tests in corefx to exercise the rare path?

Not currently checked in, as the special case didn't previously exist.  But I have a little mini-stress test locally that typically ends up hitting both of the special paths (too long in the middle, incorrect length at the end), and I'll put up a PR for it tomorrow once this change is checked in and consumable in the packages.  I don't want to put it up yet as it also includes a few more tests that don't currently pass, e.g. checked for the 0-length optimization that exists in the other string.Concat overloads, doesn't currently exist in the string.Concat(string[]) overload, but now with this change does exist in that one.

**Comment by stephentoub** [comment]
Updated.

File: src/mscorlib/src/System/String.cs [hunk 2]
```diff
@@ -3339,24 +3339,63 @@ public static String Concat(Object arg0, Object arg1, Object arg2, Object arg3,
...
-                    throw new OutOfMemoryException();
+                if (value != null)
+                {
+                    totalLengthLong += value.Length;
                 }
             }
-            
-            return ConcatArray(internalValues, totalLength);
+
+            // If it's too long, fail, or if it's empty, return an empty string.
```
Comment (jkotas) [inline src/mscorlib/src/System/String.cs]: ConcatArray is called from a single place now - can fold it into its caller.

File: src/mscorlib/src/System/String.cs [hunk 3]
```diff
@@ -3339,24 +3339,63 @@ public static String Concat(Object arg0, Object arg1, Object arg2, Object arg3,
...
+            {
+                string value = values[i];
+                if (!string.IsNullOrEmpty(value))
+                {
+                    int valueLen = value.Length;
+                    if (valueLen > totalLength - copiedLength)
+                    {
+                        // Something changed concurrently.  Fall back to doing the concat again,
+                        // this time with a defensive copy.  This should be extremely rare.
+                        return Concat((string[])values.Clone());
```
Comment (jkotas) [inline src/mscorlib/src/System/String.cs]: Nit - `copiedLength = -1; break;` would be a bit more efficient (smaller code).

**Comment by jkotas** [comment]
LGTM otherwise

File: src/mscorlib/src/System/String.cs (hunk 2 above)
Comment (stephentoub) [inline src/mscorlib/src/System/String.cs]: Fixed.

File: src/mscorlib/src/System/String.cs (hunk 3 above)
Comment (stephentoub) [inline src/mscorlib/src/System/String.cs]: Fixed.

**Comment by hughbe** [comment]
Nice. Do we have a similar situation with other `params string[]` methods such as `string.Join` or `string.Format`?

**Comment by stephentoub** [comment]
> Do we have a similar situation with other  params string[]  methods such as  string.Join  or  string.Format ?

Not as-is, in that they use StringBuilderCache and Append directly to the received StringBuilder.  The StringBuilderCache approach is a good one when it's expected that the returned string will fit within the limits of the StringBuilder cache, and it's a good approach for string.Format where there's a lot more complexity involved.

But... it might be interesting to look at string.Join specifically to see whether changing to use an approach like the one now employed by Concat would help with performance.  At the moment I'm not seeing significant perf downsides to doing so (other than the super rare case of concurrent manipulation), but there may be something I'm missing, and I expect the implementation would end up being slightly more complex, though not terribly.  For the cases where the string ends up being longer than what fits in a StringBuilder from the cache, I'd expect this approach would be more efficient.

**Comment by ellismg** [comment]
LGTM as well.  Thanks, Steve.

**Comment by stephentoub** [comment]
Test Windows_NT x64 Debug Build and Test please
Test Windows_NT x64 Release Priority 1 Build and Test please

**Comment by stephentoub** [comment]
Not sure why the Windows_NT x86 Debug Legacy Backend Build and Test leg is still showing as yellow; clicking on the Details link shows that it completed successfully.

**Pull request merged**